internal/pkg/manifest/union.go

```diff
@@ -0,0 +1,163 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"gopkg.in/yaml.v3"
+)
+
+// Union is a type used for yaml keys that may be of type A or B.
+// Union will only ever hold one of the underlying types, never both.
+//
+// Use NewUnionA() and NewUnionB() to create a Union with an underlying
+// type already set. See Unmarshal() for details on how yaml is decoded
+// into Union.
+//
+// Union is exported to enable type embedding.
+type Union[A, B any] struct {
+	// isA is true if the underlying type of Union is A.
+	isA bool
+
+	// A holds the value of Union if IsA() is true.
+	// If IsA() is false, this is the zero value of type A.
+	//
+	// A is exported to support mergo. It should not be set
+	// directly. Use NewUnionA() to create a Union with A set.
+	A A
+
+	// isB is true if the underlying type of Union is B.
+	isB bool
+
+	// B holds the value of Union if IsB() is true.
+	// If IsB() is false, this is the zero value of type B.
+	//
+	// B is exported to support mergo. It should not be set
+	// directly. Use NewUnionB() to create a Union with B set.
+	B B
+}
+
+// NewUnionA creates a new Union[A, B] with the underlying
+// type set to A, holding val.
+func NewUnionA[A, B any](val A) Union[A, B] {
+	return Union[A, B]{
+		isA: true,
+		A:   val,
+	}
+}
+
+// NewUnionB creates a new Union[A, B] with the underlying
+// type set to B, holding val.
+func NewUnionB[A, B any](val B) Union[A, B] {
+	return Union[A, B]{
+		isB: true,
+		B:   val,
+	}
+}
+
+// IsA returns true if the underlying value of t is type A.
+func (t *Union[_, _]) IsA() bool {
+	return t.isA
+}
+
+// IsB returns true if the underlying value of t is type B.
+func (t *Union[_, _]) IsB() bool {
+	return t.isB
+}
+
+// Value returns either the underlying value of t, which is either type
+// A or B, depending on which was filled by yaml.Unmarshal or created
+// with NewUnionA/B(). If neither have been/were set, Value returns nil.
+func (t *Union[A, B]) Value() any {
```

Comment:
Does this need to be a public func?

Reply:
good catch! i think we can actually just remove it and move it's logic to `MarshalYAML()`